src/Connection.php

```diff
@@ -154,77 +149,72 @@ public static function connect($dsn, $user = null, $password = null, $args = [])
-                ], $args));
-
-                break;
-            case 'counter': // deprecated - will be removed in dec-2020
-            case 'profile':
-                $c = new Debug\Profiler\Connection(array_merge([
-                    'connection' => static::connect($dsn['rest'], $dsn['user'], $dsn['pass']),
-                ], $args));
-
-                break;
-            // let PDO handle the rest
-            default:
-                $c = new static(array_merge([
-                    'connection' => static::connect(static::getPDO($dsn)),
-                ], $args));
+        /**
+         * @var Connection $connectionClass
+         */
+        $connectionClass = self::resolveConnection($dsn['driverType']);
+
+        return new $connectionClass(array_merge([
+            'connection' => $connectionClass::establishConnection($dsn),
+        ], $args));
+    }
+
+    /**
+     * Adds connection class to the registry for resolving in Connection::resolve method.
+     *
+     * Can be used as:
+     *
+     * Connection::register('mysql', MySQL\Connection::class), or
+     * MySQL\Connection::register()
+     *
+     * CustomDriver\Connection must be descendant of Connection class.
+     *
+     * @param string $connectionType
+     * @param string $connectionClass
+     */
+    public static function registerConnection($connectionType = null, $connectionClass = null)
+    {
+        if (!$connectionClass && is_a($connectionType, self::class, true)) {
+            $connectionClass = $connectionType;
+            $connectionType = null;
+        }
+
+        $connectionClass = $connectionClass ?? static::class;
+
+        $connectionType = $connectionType ?? $connectionClass::defaultDriverType();
+
+        if (is_array($driverTypes = $connectionType)) {
+            foreach ($driverTypes as $connectionType => $connectionClass) {
+                if (is_numeric($connectionType)) {
+                    $connectionType = $connectionClass::defaultDriverType();
+                }
+
+                static::registerConnection($connectionType, $connectionClass);
+            }
         }
 
-        return $c;
+        self::$connectionRegistry[$connectionType] = $connectionClass;
+    }
+
+    /**
+     * Resolves the connection class to use based on driver type.
+     *
+     * @param string $connectionType
+     *
+     * @return string
+     */
+    public static function resolveConnection($connectionType)
+    {
+        return self::$connectionRegistry[$connectionType] ?? static::class;
     }
 
     /**
-     * Returns new PDO object.
+     * Establishes connection based on a $dsn
+     * By default connection is established using new PDO object which can be overridden in child classes.
      *
      * This does not silence PDO errors.
      */
-    protected static function getPDO(array $dsn)
+    public static function establishConnection(array $dsn)
```

Comment:
If `connect` stays (and it should), this should never be public and what is the reason of the rename? We can name it `connectDriver()` (to be consistent with connect) and it should return `PDO` only.

The special classes should `connect` the underlaying drivers by overriding the `connect` method.

Reply:
I will make it protected and rename it to connectDriver, this is a good name.
The reason for rename is that it does not always return PDO, t may also return another Connection.

Reply:
fine with this